python/cudf/cudf/core/index.py

```diff
@@ -785,14 +785,61 @@ def difference(self, other, sort=None):
 
         return difference
 
-    def _apply_op(self, fn, other=None):
+    def _binaryop(self, other, fn, fill_value=None, reflect=False):
+        # TODO: Rather than including an allowlist of acceptable types, we
+        # should instead return NotImplemented for __all__  other types. That
+        # will allow other types to support binops with cudf objects if they so
+        # choose, and just as importantly will allow better error messages if
+        # they don't support it.
+        if isinstance(other, (cudf.DataFrame, cudf.Series)):
+            return NotImplemented
 
-        idx_series = cudf.Series(self, name=self.name)
-        op = getattr(idx_series, fn)
-        if other is not None:
-            return as_index(op(other))
-        else:
-            return as_index(op())
+        return super()._binaryop(other, fn, fill_value, reflect)
+
+    def _copy_construct(self, **kwargs):
+        # Need to override the parent behavior because pandas allows operations
+        # on unsigned types to return signed values, forcing us to choose the
+        # right index type here.
+        data = kwargs.get("data")
+        cls = self.__class__
+
+        if data is not None:
+            if self.dtype != data.dtype:
+                # TODO: This logic is largely copied from `as_index`. The two
+                # should be unified via a centralized type dispatching scheme.
+                if isinstance(data, NumericalColumn):
+                    try:
+                        cls = _dtype_to_index[data.dtype.type]
+                    except KeyError:
+                        cls = GenericIndex
+                        # TODO: GenericIndex has a different API for __new__
+                        # than other Index types. Refactoring Index types will
+                        # be necessary to clean this up.
+                        kwargs["values"] = kwargs.pop("data")
+                elif isinstance(data, StringColumn):
+                    cls = StringIndex
+                elif isinstance(data, DatetimeColumn):
+                    cls = DatetimeIndex
+                elif isinstance(data, TimeDeltaColumn):
+                    cls = TimedeltaIndex
+                elif isinstance(data, CategoricalColumn):
+                    cls = CategoricalIndex
+            elif cls is RangeIndex:
+                # RangeIndex must convert to other numerical types for ops
+
+                # TODO: The one exception is that multiplication of a
+                # RangeIndex in pandas results in another RangeIndex.
+                # Propagating that information through cudf with the current
+                # APIs is likely more hacky than it's worth, and it's more a
+                # performance loss than a functional loss to have an Int64Index
+                # instead of a RangeIndex. As Index logic is cleaned up it may
+                # become more feasible to ensure the appropriate type.
```

Comment:
It's not true that it's only a performance loss. Someone could have something like `(RangeIndex(0, 5) * 5).stop` which would then be broken.

Reply:
That's a good point. This behavior is also not a regression, though, it's the current behavior on `branch-0.20`:
```
>>> import cudf
>>> cudf.RangeIndex(0, 5)
RangeIndex(start=0, stop=5, step=1)
>>> cudf.RangeIndex(0, 5)*5
Int64Index([0, 5, 10, 15, 20], dtype='int64')
```

That being the case, are you OK with me just updating this comment for now but pushing on the fix until we can do a cleaner refactoring of the `Index` classes?

Reply:
Sounds good!